{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-11-06T00:57:53.157021Z",
     "start_time": "2025-11-06T00:57:53.155431Z"
    }
   },
   "source": [
    "import csv\n",
    "import glob\n",
    "import re\n",
    "import subprocess"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T00:57:53.177306Z",
     "start_time": "2025-11-06T00:57:53.174989Z"
    }
   },
   "cell_type": "code",
   "source": "from collections import defaultdict",
   "id": "7efbbe3c4ff3d7c2",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T00:57:55.095880Z",
     "start_time": "2025-11-06T00:57:53.225437Z"
    }
   },
   "cell_type": "code",
   "source": "from pybedtools import BedTool",
   "id": "7a970abe4ce319ff",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T01:06:55.536602Z",
     "start_time": "2025-11-06T00:57:55.177615Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for path in glob.iglob(\"data/ref/*/annot/*.gtf.gz\"):\n",
    "    name = re.search('(?<=ref/)[^/]+', path).group(0)\n",
    "\n",
    "    save_path = f\"data/ref/{name}/annot/out/meta_exon.csv\"\n",
    "    subprocess.call(f\"python main/merge.py -g {path} -o {save_path}\", shell=True)"
   ],
   "id": "223a9db300e5d9e1",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T01:06:55.630127Z",
     "start_time": "2025-11-06T01:06:55.622589Z"
    }
   },
   "cell_type": "code",
   "source": "name2hsap = defaultdict(str)",
   "id": "92a92d0654d60b38",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T01:56:55.477922Z",
     "start_time": "2025-11-06T01:06:55.638573Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for fp in glob.iglob(\"data/homology/exons/chrom/*.csv\"):\n",
    "    with open(fp, 'r') as f:\n",
    "        reader = csv.DictReader(f)\n",
    "        \n",
    "        for row in reader:\n",
    "            hsap = row.pop('homo_sapiens')\n",
    "            \n",
    "            if hsap == \"-\":\n",
    "                # `row['anchor']`, the H. sap. meta-exon, has no alignment.\n",
    "                continue\n",
    "                \n",
    "            for name, pos in row.items():\n",
    "                if name == 'anchor' or pos == \"-\":\n",
    "                    continue\n",
    "\n",
    "                chrom, beg, end, strand = re.split('[:-]', pos)\n",
    "                strand = \"+\" if int(strand) else \"-\"\n",
    "                \n",
    "                name2hsap[name] += (\n",
    "                    f\"{chrom}\\t\"\n",
    "                    f\"{beg}\\t\"\n",
    "                    f\"{end}\\t\"\n",
    "                    f\"{hsap}\\t\"\n",
    "                    f\"0\\t\"\n",
    "                    f\"{strand}\\n\"\n",
    "                )"
   ],
   "id": "5a331fb5411bed4f",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T02:42:45.906315Z",
     "start_time": "2025-11-06T01:57:02.921356Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for fp in glob.iglob(\"data/ref/*/annot/out/meta_exon.csv\"):\n",
    "    name = re.search(\"(?<=ref/)[^/]+\", fp).group(0)\n",
    "    exon_bed = str()\n",
    "\n",
    "    with open(fp, 'r') as f:\n",
    "        reader = csv.reader(f)\n",
    "        \n",
    "        for row in reader:\n",
    "            gene, gene_func, chrom, meta_beg, meta_end, strand, meta_kind, make_up = row\n",
    "\n",
    "            bin_strand = int(strand == \"+\")\n",
    "            meta_exon = f\"{meta_kind}={chrom}:{meta_beg}-{meta_end}:{bin_strand}\"\n",
    "            \n",
    "            exon_bed += (\n",
    "                f\"{chrom}\\t\"\n",
    "                f\"{meta_beg}\\t\"\n",
    "                f\"{meta_end}\\t\"\n",
    "                f\"{gene_func}\\t\"\n",
    "                f\"{meta_exon}\\t\"\n",
    "                f\"{strand}\\n\"\n",
    "            )\n",
    "\n",
    "    hsap_bed = BedTool(name2hsap[name], from_string=True)\n",
    "    exon_bed = BedTool(exon_bed, from_string=True)\n",
    "    \n",
    "    hsap_exon_bed = hsap_bed.intersect(exon_bed, wao=True)\n",
    "    hsap_exon_bed.saveas(f\"data/ref/{name}/annot/out/hsap_exon_homology.bed\")"
   ],
   "id": "21e771201f2e76df",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T02:42:46.023006Z",
     "start_time": "2025-11-06T02:42:46.002915Z"
    }
   },
   "cell_type": "code",
   "source": [
    "exon_map = defaultdict(dict)\n",
    "names = list()"
   ],
   "id": "591c81b47f4236e7",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T02:42:46.038481Z",
     "start_time": "2025-11-06T02:42:46.031567Z"
    }
   },
   "cell_type": "code",
   "source": [
    "hsap_exon_data = dict()\n",
    "hsap_gene_data = dict()"
   ],
   "id": "6be6a21604e50f1e",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T02:42:46.708565Z",
     "start_time": "2025-11-06T02:42:46.045981Z"
    }
   },
   "cell_type": "code",
   "source": [
    "with open(\"../data/ref/homo_sapiens/annot/out/meta_exon.csv\", 'r') as f:\n",
    "    reader = csv.reader(f)\n",
    "    next(reader)\n",
    "\n",
    "    for row in reader:\n",
    "        gene, gene_func, chrom, beg, end, strand, exon_type, make_up = row\n",
    "\n",
    "        strand = int(strand == \"+\")\n",
    "        meta_exon = f\"{chrom}:{beg}-{end}:{strand}\"\n",
    "\n",
    "        hsap_exon_data[meta_exon] = exon_type\n",
    "        hsap_gene_data[meta_exon] = gene, gene_func"
   ],
   "id": "67cb2fe588e04e3c",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T02:43:14.637853Z",
     "start_time": "2025-11-06T02:42:46.717514Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for fp in glob.iglob(\"data/ref/*/annot/out/hsap_exon_homology.bed\"):\n",
    "    name = re.search(\"(?<=ref/)[^/]+\", fp).group(0)\n",
    "\n",
    "    if name == \"homo_sapiens\":\n",
    "        continue\n",
    "\n",
    "    # I want H. sap. to be column 1, so I add it to the header\n",
    "    # manually.\n",
    "    names.append(name)\n",
    "\n",
    "    with open(fp, 'r') as f:\n",
    "        reader = csv.reader(f, delimiter=\"\\t\")\n",
    "\n",
    "        for row in reader:\n",
    "            overlap = int(row.pop())\n",
    "\n",
    "            h_chrom, *h_pos, h_exon, _, h_strand = row[:6]\n",
    "            o_chrom, *o_pos, _, o_exon, o_strand = row[6:]\n",
    "\n",
    "            if overlap:\n",
    "                exon_map[h_exon][name] = o_exon\n",
    "            else:\n",
    "                exon_map[h_exon][name] = \"-\""
   ],
   "id": "9441eb921bd2adaa",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T02:43:36.104073Z",
     "start_time": "2025-11-06T02:43:31.307988Z"
    }
   },
   "cell_type": "code",
   "source": [
    "with open(\"../data/homology/exons/exons.csv\", 'w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow([\"hsap_exon\", \"hsap_gene\", \"hsap_gene_biotype\"] + names)\n",
    "\n",
    "    # `exon_map` is populated from the `bedtools` output, so if an exon\n",
    "    # has no alignment in any species, it's excluded.\n",
    "    for h_exon in exon_map:\n",
    "        h_type = hsap_exon_data[h_exon]\n",
    "        h_gene, h_func = hsap_gene_data[h_exon]\n",
    "\n",
    "        row = [\n",
    "            h_type + \"=\" + h_exon,\n",
    "            h_gene,\n",
    "            h_func,\n",
    "        ]\n",
    "\n",
    "        for name in names:\n",
    "            exon = exon_map[h_exon].get(name, \"-\")\n",
    "            row.append(exon)\n",
    "\n",
    "        writer.writerow(row)"
   ],
   "id": "d6897ba883271e26",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-06T02:43:36.113633Z",
     "start_time": "2025-11-06T02:43:36.111542Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "2fdcaf06f3c61909",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
